public/javascripts/SVValidate/src/panorama/GSVOverlay.js

```diff
@@ -0,0 +1,143 @@
+/**
+ * An additional layer on top of the Google StreetView object
+ * on validation interface. This layer handles panning.
+ *
+ */
+function GSVOverlay () {
+    let self = this;
+    var panningDisabled = false;
+    var viewControlLayer = $("#viewControlLayer");
+
+    // Mouse status and mouse event callback functions
+    var mouseStatus = {
+        currX: 0,
+        currY: 0,
+        prevX: 0,
+        prevY: 0,
+        leftDownX: 0,
+        leftDownY: 0,
+        leftUpX: 0,
+        leftUpY: 0,
+        isLeftDown: false
+    };
+
+    /**
+     * Disables panning on the GSV window.
+     */
+    function disablePanning() {
+        panningDisabled = true;
+    }
+
+    /**
+     * Enables panning on the GSV window.
+     */
+    function enablePanning() {
+        panningDisabled = false;
+    }
+
+    /**
+     * This is a callback function that is fired with the mouse down event
+     * on the view control layer (where you control street view angle.)
+     * @param e
+     */
+    function handlerViewControlLayerMouseDown (e) {
+        mouseStatus.isLeftDown = true;
+        mouseStatus.leftDownX = mouseposition(e, this).x;
+        mouseStatus.leftDownY = mouseposition(e, this).y;
+        viewControlLayer.css("cursor", "url(/assets/javascripts/SVLabel/img/cursors/closedhand.cur) 4 4, move");
+
+        //This is necessary for supporting touch devices, because there is no mouse hover
+        mouseStatus.prevX = mouseposition(e, this).x;
+        mouseStatus.prevY = mouseposition(e, this).y;
+    }
+
+    /**
+     * This is a callback function that is called with mouse up event on
+     * the view control layer (where you change the Google Street view angle.
+     * @param e
+     */
+    function handlerViewControlLayerMouseUp (e) {
+        viewControlLayer.css("cursor", "url(/assets/javascripts/SVLabel/img/cursors/openhand.cur) 4 4, move");
+        mouseStatus.isLeftDown = false;
+        mouseStatus.leftUpX = mouseposition(e, this).x;
+        mouseStatus.leftUpY = mouseposition(e, this).y;
+
+    }
+    /**
+     * Handles mouse leaving control view.
+     * @param e
+     */
+    function handlerViewControlLayerMouseLeave (e) {
+        viewControlLayer.css("cursor", "url(/assets/javascripts/SVLabel/img/cursors/openhand.cur) 4 4, move");
+        mouseStatus.isLeftDown = false;
+    }
+
+    /**
+     * This is a callback function that is fired when a user moves a mouse on the
+     * view control layer where you change the pov.
+     */
+    function handlerViewControlLayerMouseMove (e) {
+        mouseStatus.currX = mouseposition(e, this).x;
+        mouseStatus.currY = mouseposition(e, this).y;
+
+        var timestamp = new Date().getTime();  // waits till the pano is fully loaded
+        if ((timestamp - svv.panorama.getProperty("prevSetPanoTimestamp") > 800) && mouseStatus.isLeftDown && panningDisabled === false) {
```

Comment:
What is this doing exactly? Are we waiting 800ms and assuming the pano has fully loaded at that point?

Reply:
There is a line of code in Keyboard.js file that disallows the user to validate for a period of 800ms after the new pano is set. I believe this value was experimentally determined by the person who wrote that file. So, what the code above does is it ensures that the user can't pan during those first 800ms.

Reply:
I believe that the 800 ms delay was actually added to prevent users from speed-auditing without actually looking at the imagery. I might be wrong about that, but I thought that is what it was for. If that is the case, we might be able to bring down that delay in this case so that users can pan sooner.